### PR TITLE
Support Longchat

### LINKED ISSUE #333: Support for Condensed RotaryEmbeddings
looks like there are various methods of extending context length, suck as superhot, ntk-aware, and condensed rope. This request is to track support for condensed rotaryembeddings as it seems to have the best performance at long contexts atm.

https://lmsys.org/blog/2023-06-29-longchat/
https://github.com/lm-sys/FastChat/blob/3f0c6e54498e179098ead9a596929e23327ad75c/fastchat/model/llama_condense_monkey_patch.py#L68

### PR DESCRIPTION
**Pull request by LiuXiaoxuanPKU**
Add LlamaLinearScalingRotaryEmbedding, LlamaDynamicNTKScalingRotaryEmbedding. Attempt to fix #333, #464, #479

### CODE REVIEW

**Comment by zhuohan123** [comment]
CC @DachengLi1

**Comment by DachengLi1** [comment]
Wonderful! Thanks a lot Lily and Zhuohan! Will this be merged soon?

**Comment by zhuohan123** [comment]
> Wonderful! Thanks a lot Lily and Zhuohan! Will this be merged soon?

I believe so! @LiuXiaoxuanPKU is working on some correctness tests and making sure everything works for a long context. Feel free to try out this PR if you would like to start immediately!

**Comment by winglian** [comment]
@LiuXiaoxuanPKU You also need to add rope_scaling as an argument to argparse (https://github.com/LiuXiaoxuanPKU/vllm/blob/longchat/vllm/engine/arg_utils.py#L40) , otherwise this call on line 141 failse
`engine_args = cls(**{attr: getattr(args, attr) for attr in attrs})`

**Comment by lucasjinreal** [comment]
@LiuXiaoxuanPKU Will also support Baichuan model?

https://github.com/keezen/ntk_alibi like using AlibiNTK for scaling in alibi

**Comment by desperadoola** [comment]
https://github.com/vllm-project/vllm/blob/66c54aa9c33555a6b41421d57d3ad6c1bf004ec9/vllm/model_executor/models/baichuan.py#L150-L156

Seems like we never tell `PagedAttentionWithRoPE` the original model `position_embedding`? The default position_embedding in PagedAttentionWithRoPE is 8192, during register_buffer it seems to register a sequence longer than the  original model?

**Comment by alanxmay** [comment]
🤔 Is there any reason to prevent this PR from being merged

**Comment by LiuXiaoxuanPKU** [comment]
> 🤔 Is there any reason to prevent this PR from being merged

We think vLLM might have some correctness issues. It might or might not be caused by this PR. To be more concrete, take a look at `tests/test_longprompt.py` in this PR. The four test cases should pass (HF passes all four cases), but currently, vLLM fails on them. We are still debugging it.

**Comment by lucasjinreal** [comment]
@LiuXiaoxuanPKU  does it happens on only longchat or normal inferences compare with hf? I actually always some issue compare with hf even the params are same.

**Comment by alanxmay** [comment]
@LiuXiaoxuanPKU Appreciate your work!

**Comment by KanyiTang** [comment]
Testing on an A100 80GB environment, encountering a CUDA error "invalid argument" when the context length exceeds 13k. What should I do?

**Comment by LiuXiaoxuanPKU** [comment]
> @LiuXiaoxuanPKU does it happens on only longchat or normal inferences compare with hf? I actually always some issue compare with hf even the params are same.

Sorry for the late reply. vLLM uses different kernels compared with HF. When vllm's output is different from HF's, it's very difficult to tell if it's a numerical problem or an actual bug. We might only be able to say vllm has bugs when vllm consistently performs worse (e.g., lower accuracy) on some tasks/benchmarks. Therefore, if anyone has such tasks that vllm can reproduce and look into, it will be highly appreciated. The team is actively looking for such cases/examples, if you believe some single cases are wrong, please share as well!

**Comment by idealover** [comment]
> > @LiuXiaoxuanPKU does it happens on only longchat or normal inferences compare with hf? I actually always some issue compare with hf even the params are same.
> 
> Sorry for the late reply. vLLM uses different kernels compared with HF. When vllm's output is different from HF's, it's very difficult to tell if it's a numerical problem or an actual bug. We might only be able to say vllm has bugs when vllm consistently performs worse (e.g., lower accuracy) on some tasks/benchmarks. Therefore, if anyone has such tasks that vllm can reproduce and look into, it will be highly appreciated. The team is actively looking for such cases/examples, if you believe some single cases are wrong, please share as well!

I have been using your patch for rope-scaling for a while and I didn't face any issue until now. I tried with Llama 70b 8k and 32k with no weird outputs and the benchmarks were as expected as well.

File: vllm/config.py [hunk 1]
```diff
@@ -55,6 +61,7 @@ def __init__(
         load_format: str,
         dtype: str,
         seed: int,
+        rope_sclaing: Optional[dict],
```
Comment (orangetin) [inline vllm/config.py:64]: rope_scaling is misspelled here and on line 77

**Comment by mahaddad** [comment]
> > > @LiuXiaoxuanPKU does it happens on only longchat or normal inferences compare with hf? I actually always some issue compare with hf even the params are same.
> > 
> > 
> > Sorry for the late reply. vLLM uses different kernels compared with HF. When vllm's output is different from HF's, it's very difficult to tell if it's a numerical problem or an actual bug. We might only be able to say vllm has bugs when vllm consistently performs worse (e.g., lower accuracy) on some tasks/benchmarks. Therefore, if anyone has such tasks that vllm can reproduce and look into, it will be highly appreciated. The team is actively looking for such cases/examples, if you believe some single cases are wrong, please share as well!
> 
> I have been using your patch for rope-scaling for a while and I didn't face any issue until now. I tried with Llama 70b 8k and 32k with no weird outputs and the benchmarks were as expected as well.



> > > @LiuXiaoxuanPKU does it happens on only longchat or normal inferences compare with hf? I actually always some issue compare with hf even the params are same.
> > 
> > 
> > Sorry for the late reply. vLLM uses different kernels compared with HF. When vllm's output is different from HF's, it's very difficult to tell if it's a numerical problem or an actual bug. We might only be able to say vllm has bugs when vllm consistently performs worse (e.g., lower accuracy) on some tasks/benchmarks. Therefore, if anyone has such tasks that vllm can reproduce and look into, it will be highly appreciated. The team is actively looking for such cases/examples, if you believe some single cases are wrong, please share as well!
> 
> I have been using your patch for rope-scaling for a while and I didn't face any issue until now. I tried with Llama 70b 8k and 32k with no weird outputs and the benchmarks were as expected as well.

@idealover are you able to share what GPU setup and parameters you ran the server with to get the 32K context window to run? I am coming from this [issue](https://github.com/vllm-project/vllm/issues/905) and considering testing out this PR to see if it works for me as well.

**Comment by idealover** [comment]
> > > > @LiuXiaoxuanPKU does it happens on only longchat or normal inferences compare with hf? I actually always some issue compare with hf even the params are same.
> > > 
> > > 
> > > Sorry for the late reply. vLLM uses different kernels compared with HF. When vllm's output is different from HF's, it's very difficult to tell if it's a numerical problem or an actual bug. We might only be able to say vllm has bugs when vllm consistently performs worse (e.g., lower accuracy) on some tasks/benchmarks. Therefore, if anyone has such tasks that vllm can reproduce and look into, it will be highly appreciated. The team is actively looking for such cases/examples, if you believe some single cases are wrong, please share as well!
> > 
> > 
> > I have been using your patch for rope-scaling for a while and I didn't face any issue until now. I tried with Llama 70b 8k and 32k with no weird outputs and the benchmarks were as expected as well.
> 
> > > > @LiuXiaoxuanPKU does it happens on only longchat or normal inferences compare with hf? I actually always some issue compare with hf even the params are same.
> > > 
> > > 
> > > Sorry for the late reply. vLLM uses different kernels compared with HF. When vllm's output is different from HF's, it's very difficult to tell if it's a numerical problem or an actual bug. We might only be able to say vllm has bugs when vllm consistently performs worse (e.g., lower accuracy) on some tasks/benchmarks. Therefore, if anyone has such tasks that vllm can reproduce and look into, it will be highly appreciated. The team is actively looking for such cases/examples, if you believe some single cases are wrong, please share as well!
> > 
> > 
> > I have been using your patch for rope-scaling for a while and I didn't face any issue until now. I tried with Llama 70b 8k and 32k with no weird outputs and the benchmarks were as expected as well.
> 
> @idealover are you able to share what GPU setup and parameters you ran the server with to get the 32K context window to run? I am coming from this [issue](https://github.com/vllm-project/vllm/issues/905) and considering testing out this PR to see if it works for me as well.

I ran it on 4x A100 80GBs with bf16 (its a Llama 2 70b finetune) and I only mention these additional parameters when using the vllm - `--tensor-parallel-size 4 --rope-scaling '{"factor":8.0, "type":"linear"}`

**Comment by alanxmay** [comment]
@LiuXiaoxuanPKU I'm curious why remove nth?

**Review by WoosukKwon — APPROVED** [review]
@LiuXiaoxuanPKU Awesome! Many thanks for the great work and sorry again for the very late review.

I've made several changes (mostly on the code style):
1. I temporarily removed the tests for faster integration. I will submit another PR to add test for RoPE scaling.
2. As we discussed offline, I removed `rope_scaling` from `ModelConfig` and `EngineArgs`. Now `rope_scaling` is always read from the model's `config.json`.
3. I refactored `rotary_embedding.py` and added `DynamicNTKScalingRotaryEmbedding`.

**Comment by hahmad2008** [comment]
@WoosukKwon @LiuXiaoxuanPKU  Can i extend model llama3.1-8b in inference time by passing these params?
`--rope-scaling and --rope-theta`
So if the model was trained on 8k, then what should I pass to support 128k length?

**Pull request merged**